File: 597-subtree-with-maximum-average/dfs.cc
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class ResultType {
public:
    int size;
    int sum;
    //ResultType() : size(0), sum(0) {}
    ResultType(int size, int sum) {
        this->size = size;
        this->sum = sum;
    }
 };

class Solution {
public:
    /**
     * @param root: the root of binary tree
     * @return: the root of the maximum average of subtree
     */
    TreeNode * findSubtree2(TreeNode * root) {
        // write your code here
        helper(root);
        return node;
    }
    ResultType helper(TreeNode* root) {
        if (root == NULL) {
            return ResultType(0, 0);
        }
        ResultType left = helper(root->left);
        ResultType right = helper(root->right);
        ResultType result = ResultType(left.size + right.size + 1, left.sum + right.sum + root->val);
        if (node == NULL || result.sum * data.size > data.sum * result.size) {
            node = root;
            data = result;
        }
        return result;
    }
private:
    TreeNode* node = NULL;
    ResultType data = ResultType(0, 0);
};

File: 70-binary-tree-level-order-traversal-ii/dfs.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */

public class Solution {
    /**
     * @param root: A tree
     * @return: buttom-up level order a list of lists of integer
     */
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        // write your code here
        List<List<Integer>> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        dfs(root, 0, result);
        return result;
    }
    void dfs(TreeNode node, int level, List<List<Integer>> result) {
        if (node == null) {
            return;
        }
        if (level >= result.size()) {
            result.add(0, new ArrayList<>());
        }
        dfs(node.left, level + 1, result);
        dfs(node.right, level + 1, result);
        result.get(result.size() - level - 1).add(node.val);
    }
}
